dll/win32/wdmaud.drv/legacy.c

```diff
@@ -1,89 +1,306 @@
+
+                    /* Wait for I/O complete */
+                    WaitForSingleObject(HeaderExtension->Overlapped->hEvent, INFINITE);
+
+                    /* Enter critical section */
+                    EnterCriticalSection(&WdmAudCriticalSection);
+
+                    DPRINT("4\n");
+                    if (DeviceInfo)
+                    {
+                        /* Complete current header */
+                        CompleteWaveHeader(DeviceInfo);
+                    }
+                    else
+                    {
+                        DPRINT1("Invalid device info data %p\n", DeviceInfo);
+                        break;
+                    }
+                }
+            }
+            /* Leave critical section */
+            LeaveCriticalSection(&WdmAudCriticalSection);
+        }
+        else
+        {
+            DPRINT("6\n");
+            if (DeviceInfo->DeviceState.bStart)
+            {
+                DPRINT("7\n");
+#ifdef USE_MMIXER_LIB
+
+                /* make sure the pin is stopped */
+                MMixerSetWaveStatus(&MixerContext, DeviceInfo->hDevice, KSSTATE_PAUSE);
+                MMixerSetWaveStatus(&MixerContext, DeviceInfo->hDevice, KSSTATE_ACQUIRE);
+                MMixerSetWaveStatus(&MixerContext, DeviceInfo->hDevice, KSSTATE_STOP);
+
+#else
+                /* Stop streaming if no more data to stream */
+                Result = WdmAudIoControl(DeviceInfo,
+                                         0,
+                                         NULL,
+                                         DeviceInfo->DeviceType == WAVE_OUT_DEVICE_TYPE ?
+                                         IOCTL_PAUSE_PLAYBACK : IOCTL_PAUSE_CAPTURE);
+                if (!MMSUCCESS(Result))
+                {
+                    DPRINT1("Call to %ls failed with %d\n",
+                            DeviceInfo->DeviceType == WAVE_OUT_DEVICE_TYPE ?
+                            L"IOCTL_PAUSE_PLAYBACK" : L"IOCTL_PAUSE_CAPTURE",
+                            GetLastError());
+                    break;
+                }
+#endif
+                DeviceInfo->DeviceState.bStart = FALSE;
+		    }
 
-    /* setup wait objects */
-    WaitObjects[0] = Instance->hNotifyEvent;
-    WaitObjects[1] = Instance->hStopEvent;
+            /* Leave critical section */
+            LeaveCriticalSection(&WdmAudCriticalSection);
 
-    /* zero device info */
-    ZeroMemory(&DeviceInfo, sizeof(WDMAUD_DEVICE_INFO));
+            /* Wait for stop event complete */
+            WaitForSingleObject(DeviceInfo->DeviceState.hNotifyEvent, INFINITE);
 
-    DeviceInfo.hDevice = Instance->Handle;
-    DeviceInfo.DeviceType = MIXER_DEVICE_TYPE;
+            /* Done */
+            break;
+        }
+    }
 
-    do
+    CloseHandle(DeviceInfo->DeviceState.hNotifyEvent);
+    DeviceInfo->DeviceState.hNotifyEvent = NULL;
+    DeviceInfo->DeviceState.bStartInThread = FALSE;
+    SetEvent(DeviceInfo->DeviceState.hStopEvent);
```

Comment:
This event seems misleading/unnecessary. You generally want to wait on the thread handle instead, which will signal that the thread actually stopped running.